x-pack/platform/plugins/shared/observability_ai_assistant/server/service/anonymization/index.ts

```diff
@@ -0,0 +1,294 @@
+      `Detected ${nerEntities.length} NER entities and ${regexEntities.length} regex entities, ${deduped.length} after deduplication`
+    );
+
+    return deduped;
+  }
+
+  /**
+   * New user messages are anonymised (NER + regex) and their entities stored.
+   * Assistant messages have any {hash} placeholders replaced with the original values.
+   *
+   * The function keeps a running `hashMap` built from all previously detected
+   * entities so that assistant placeholders originating from earlier user
+   * messages can be restored.
+   *
+   * If a message already has `detected_entities` it is skipped.
+   *
+   * @param messages - Messages to process for anonymization/deanonymization
+   * @returns Object containing the processed messages
+   */
+  async processMessages(messages: Message[]): Promise<{ anonymizedMessages: Message[] }> {
+    this.logger.debug(
+      `Processing ${messages.length} messages for entity detection and deanonymization`
+    );
+
+    if (!this.rules.length) {
+      return { anonymizedMessages: messages };
+    }
+
+    // Initialize hash map from existing entities
+    const hashMap = buildDetectedEntitiesMap(messages);
+
+    // Process each message and collect results
+    const result: Message[] = [];
+
+    for (const message of messages) {
+      // Skip messages that already have entities detected
+      if (message.message.detected_entities) {
+        result.push(message);
+        continue;
+      }
+
+      const { role, content } = message.message;
+
+      // Process user messages - detect entities
+      if (role === 'user' && content) {
+        try {
+          const entities = await this.detectEntities(content);
+          const processedMessage = {
+            ...message,
+            message: {
+              ...message.message,
+              detected_entities: entities.map((ent) => ({
+                entity: ent.entity,
+                class_name: ent.class_name,
+                start_pos: ent.start_pos,
+                end_pos: ent.end_pos,
+                type: ent.type,
+                hash: ent.hash,
+              })),
+            },
+          };
+
+          // Update hash map with newly detected entities
+          entities.forEach((entity) => {
+            hashMap.set(entity.hash, {
+              value: entity.entity,
+              class_name: entity.class_name,
+              type: entity.type,
+            });
+          });
+
+          result.push(processedMessage);
+        } catch (error) {
+          this.logger.error(
+            new Error('Entity detection failed for user message', { cause: error })
+          );
+          // Add the original message without entities detected
+          result.push(message);
+        }
+      }
+      // Process assistant messages - replace hash placeholders
+      else if (role === 'assistant') {
+        result.push(this.processAssistantMessage(message, hashMap));
+      } else {
+        result.push(message);
+      }
```

Comment:
we should apply this to all messages, including the system message (user-defined content) and assistant messages (which might be injected by tools)